dot/sync/message.go

```diff
@@ -4,15 +4,48 @@
 package sync
 
 import (
+	"bytes"
 	"fmt"
 
 	"github.com/ChainSafe/gossamer/dot/network"
+	"github.com/ChainSafe/gossamer/dot/peerset"
 	"github.com/ChainSafe/gossamer/dot/types"
 	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/libp2p/go-libp2p/core/peer"
 )
 
+const maxNumberOfSameRequestPerPeer uint = 2
+
 // CreateBlockResponse creates a block response message from a block request message
-func (s *Service) CreateBlockResponse(req *network.BlockRequestMessage) (*network.BlockResponseMessage, error) {
+func (s *Service) CreateBlockResponse(from peer.ID, req *network.BlockRequestMessage) (
+	*network.BlockResponseMessage, error) {
+	logger.Debugf("sync request from %s: %s", from, req.String())
+
+	encodedRequest, err := req.Encode()
+	if err != nil {
+		return nil, fmt.Errorf("encoding request: %w", err)
+	}
+
+	encodedKey := bytes.Join([][]byte{[]byte(from.String()), encodedRequest}, nil)
+	requestHash, err := common.Blake2bHash(encodedKey)
+	if err != nil {
+		return nil, fmt.Errorf("hashing encoded block request sync message: %w", err)
+	}
+
+	numOfRequests := s.seenBlockSyncRequests.Get(requestHash)
+
+	if numOfRequests > maxNumberOfSameRequestPerPeer {
```

Comment:
```suggestion
	if numOfRequests + 1 > maxNumberOfSameRequestPerPeer {
```
Shouldn't we increment this now to report the peer and exit early instead of waiting for the 3rd message spam to do this?

Reply:
but then Gossamer will punish w/o reason, I'm mean, if someone reach the limit that is not a bad thing so should not be punished, however if it goes beyond the limit then it should be punished and have its reputation decreased. That's why we should wait for the third message imo